File: src/engine/allocate.ts
import type { ClaimId } from '../core/ids.js';
import { compareClaims, required, type Claim } from '../domain/claim.js';
import { isProjectable, projectedAmount, type IncomeEvent } from '../domain/income.js';
import { isMandatory, PRIORITY, type Priority } from '../domain/priority.js';
import { clampAtZero, isZero, min, sub, sum, zero, type Money } from '../money/money.js';
import { isAfter, isOnOrBefore, isOverdue, type LocalDate } from '../time/clock.js';

export interface Allocation {
  readonly claimId: ClaimId;
  readonly priority: Priority;
  readonly label: string;
  /** What the claim needed from current liquidity in this snapshot. */
  readonly requiredNow: Money;
  readonly allocated: Money;
  readonly shortfall: Money;
}

export interface AllocationInput {
  readonly currency: string;
  readonly liquidity: Money;
  readonly claims: readonly Claim[];
  readonly today: LocalDate;
  /** Normally the next expected income event (§4). */
  readonly decisionHorizonEnd: LocalDate;
  readonly incomeEvents?: readonly IncomeEvent[];
}

export interface AllocationResult {
  readonly allocations: readonly Allocation[];
  readonly allocatedBeforeDiscretionary: Money;
  readonly protectedTotal: Money;
  readonly mandatoryFundingGap: Money;
  readonly bufferShortfall: Money;
  readonly flexibleShortfall: Money;
  readonly protectionHorizonEnd: LocalDate;
  readonly extendedBeyondDecisionHorizon: boolean;
}

/**
 * How much of a claim current liquidity must protect right now.
 *
 * For a claim due within the decision horizon this is the whole remaining
 * amount. For one falling after it, the protection horizon stretches to cover
 * only the part expected income cannot meet (§4, INV-15, fixture T11) — the
 * engine looks past payday rather than letting the user spend money that a
 * hard claim already needs.
 */
function requiredFromCurrentLiquidity(
  claim: Claim,
  input: AllocationInput,
): { amount: Money; beyondHorizon: boolean } {
  const full = required(claim);
  const due = claim.dueDate;

  if (due === undefined || isOnOrBefore(due, input.decisionHorizonEnd)) {
    return { amount: full, beyondHorizon: false };
  }
  if (isZero(full)) return { amount: full, beyondHorizon: false };

  const incomingBeforeDue = sum(
    (input.incomeEvents ?? [])
      .filter((i) => isProjectable(i) && isAfter(i.expectedDate, input.today) && isOnOrBefore(i.expectedDate, due))
      .map((i) => projectedAmount(i) ?? zero(input.currency)),
    input.currency,
  );

  return { amount: clampAtZero(sub(full, incomingBeforeDue)), beyondHorizon: true };
}

/** Overdue hard claims are elevated out of their class into P1 (§11). */
function effectivePriority(claim: Claim, today: LocalDate): Priority {
  const hard =
    claim.priority === PRIORITY.P2_HARD_OBLIGATION ||
    claim.priority === PRIORITY.P3_CARD_SPEND_RESERVE;
  if (hard && claim.dueDate !== undefined && isOverdue(claim.dueDate, today)) {
    return PRIORITY.P1_OVERDUE_HARD;
  }
  return claim.priority;
}

/**
 * The deterministic allocation waterfall (§11). Given identical inputs and
 * engine version this returns identical allocations and ordering (INV-07).
 */
export function allocate(input: AllocationInput): AllocationResult {
  const code = input.currency;

  const prepared = input.claims
    .map((claim) => {
      const { amount, beyondHorizon } = requiredFromCurrentLiquidity(claim, input);
      const priority = effectivePriority(claim, input.today);
      return { claim: { ...claim, priority }, requiredNow: amount, beyondHorizon };
    })
    .sort((a, b) => compareClaims(a.claim, b.claim));

  let remainingLiquidity = input.liquidity;
  const allocations: Allocation[] = [];
  let protectionHorizonEnd = input.decisionHorizonEnd;
  let extended = false;

  for (const { claim, requiredNow, beyondHorizon } of prepared) {
    const allocated = min(requiredNow, clampAtZero(remainingLiquidity));
    remainingLiquidity = sub(remainingLiquidity, allocated);

    allocations.push({
      claimId: claim.id,
      priority: claim.priority,
      label: claim.label,
      requiredNow,
      allocated,
      shortfall: sub(requiredNow, allocated),
    });

    if (beyondHorizon && !isZero(requiredNow) && claim.dueDate !== undefined) {
      extended = true;
      if (isAfter(claim.dueDate, protectionHorizonEnd)) protectionHorizonEnd = claim.dueDate;
    }
  }

  const before = allocations.filter((a) => a.priority !== PRIORITY.P9_DISCRETIONARY);
  const allocatedBeforeDiscretionary = sum(before.map((a) => a.allocated), code);

  /**
   * mandatory_funding_gap (§13) is the unfunded part of the mandatory classes
   * P0–P5 and P7. P6 buffer and P8 flexible shortfalls are reported separately
   * and never inflate it.
   *
   * The specification states this as Σ(required mandatory) − liquidity. That
   * form agrees with every fixture but under-reports where a non-mandatory
   * class funded earlier in the waterfall — P6 buffer sits above P7 hard goals
   * — absorbs liquidity a mandatory claim then cannot reach. Summing the
   * per-claim shortfalls is equivalent on every fixture and correct in that
   * case too, so it is what the engine computes. See tests/spec-divergence.
   */
  const mandatoryFundingGap = sum(
    allocations.filter((a) => isMandatory(a.priority)).map((a) => a.shortfall),
    code,
  );

  const shortfallAt = (p: Priority): Money =>
    sum(allocations.filter((a) => a.priority === p).map((a) => a.shortfall), code);

  return {
    allocations,
    allocatedBeforeDiscretionary,
    protectedTotal: allocatedBeforeDiscretionary,
    mandatoryFundingGap,
    bufferShortfall: shortfallAt(PRIORITY.P6_BUFFER),
    flexibleShortfall: shortfallAt(PRIORITY.P8_FLEXIBLE),
    protectionHorizonEnd,
    extendedBeyondDecisionHorizon: extended,
  };
}

/** Literal form of the §13 formula, kept so tests can assert the two agree. */
export function specFormMandatoryGap(input: AllocationInput): Money {
  const requiredMandatory = sum(
    input.claims
      .filter((c) => isMandatory(effectivePriority(c, input.today)))
      .map((c) => requiredFromCurrentLiquidity(c, input).amount),
    input.currency,
  );
  return clampAtZero(sub(requiredMandatory, input.liquidity));
}

File: app/lib/engine/allocate.dart
/// The deterministic allocation waterfall (§11) and Safe-to-Spend (§13).
library;

import 'clock.dart';
import 'domain.dart';
import 'money.dart';

class Allocation {
  const Allocation({
    required this.claimId,
    required this.priority,
    required this.label,
    required this.requiredNow,
    required this.allocated,
  });

  final String claimId;
  final Priority priority;
  final String label;

  /// What the claim needed from current liquidity in this snapshot.
  final Money requiredNow;
  final Money allocated;

  Money get shortfall => requiredNow - allocated;
}

class AllocationInput {
  const AllocationInput({
    required this.currency,
    required this.liquidity,
    required this.claims,
    required this.today,
    required this.decisionHorizonEnd,
    this.incomeEvents = const [],
  });

  final String currency;
  final Money liquidity;
  final List<Claim> claims;
  final LocalDate today;
  final LocalDate decisionHorizonEnd;
  final List<IncomeEvent> incomeEvents;

  AllocationInput withLiquidity(Money value) => AllocationInput(
        currency: currency,
        liquidity: value,
        claims: claims,
        today: today,
        decisionHorizonEnd: decisionHorizonEnd,
        incomeEvents: incomeEvents,
      );
}

class AllocationResult {
  const AllocationResult({
    required this.allocations,
    required this.allocatedBeforeDiscretionary,
    required this.mandatoryFundingGap,
    required this.bufferShortfall,
    required this.flexibleShortfall,
    required this.protectionHorizonEnd,
    required this.extendedBeyondDecisionHorizon,
  });

  final List<Allocation> allocations;
  final Money allocatedBeforeDiscretionary;
  final Money mandatoryFundingGap;
  final Money bufferShortfall;
  final Money flexibleShortfall;
  final LocalDate protectionHorizonEnd;
  final bool extendedBeyondDecisionHorizon;

  Money get protectedTotal => allocatedBeforeDiscretionary;
}

class _Prepared {
  const _Prepared(this.claim, this.requiredNow, this.beyondHorizon);
  final Claim claim;
  final Money requiredNow;
  final bool beyondHorizon;
}

/// How much of a claim current liquidity must protect right now.
///
/// A claim due inside the decision horizon needs its whole remaining amount.
/// One falling after it stretches the protection horizon to cover only the
/// part expected income cannot meet (§4, INV-15, fixture T11) — the engine
/// looks past payday rather than letting money a hard claim already needs be
/// spent today.
_Prepared _prepare(Claim claim, AllocationInput input) {
  final full = claim.required;
  final due = claim.dueDate;

  if (due == null || due <= input.decisionHorizonEnd || full.isZero) {
    return _Prepared(claim, full, false);
  }

  final incomingBeforeDue = Money.sum(
    input.incomeEvents
        .where((i) =>
            i.isProjectable && i.expectedDate > input.today && i.expectedDate <= due,)
        .map((i) => i.projectedAmount ?? Money.zero(input.currency)),
    input.currency,
  );

  return _Prepared(claim, (full - incomingBeforeDue).clampedAtZero, true);
}

/// Overdue hard claims are elevated out of their class into P1 (§11).
Priority _effectivePriority(Claim claim, LocalDate today) {
  final due = claim.dueDate;
  final isHard = claim.priority == Priority.p2HardObligation ||
      claim.priority == Priority.p3CardSpendReserve;
  if (isHard && due != null && isOverdue(due, today)) return Priority.p1OverdueHard;
  return claim.priority;
}

/// Given identical inputs and engine version this returns identical
/// allocations and ordering (INV-07).
AllocationResult allocate(AllocationInput input) {
  final currency = input.currency;

  final prepared = input.claims.map((c) {
    final p = _prepare(c, input);
    return _Prepared(
      c.copyWith(priority: _effectivePriority(c, input.today)),
      p.requiredNow,
      p.beyondHorizon,
    );
  }).toList()
    ..sort((a, b) => Claim.compare(a.claim, b.claim));

  var remaining = input.liquidity;
  final allocations = <Allocation>[];
  var protectionHorizonEnd = input.decisionHorizonEnd;
  var extended = false;

  for (final p in prepared) {
    final allocated = Money.min(p.requiredNow, remaining.clampedAtZero);
    remaining = remaining - allocated;

    allocations.add(Allocation(
      claimId: p.claim.id,
      priority: p.claim.priority,
      label: p.claim.label,
      requiredNow: p.requiredNow,
      allocated: allocated,
    ),);

    final due = p.claim.dueDate;
    if (p.beyondHorizon && !p.requiredNow.isZero && due != null) {
      extended = true;
      if (due > protectionHorizonEnd) protectionHorizonEnd = due;
    }
  }

  Money shortfallWhere(bool Function(Allocation) test) => Money.sum(
        allocations.where(test).map((a) => a.shortfall),
        currency,
      );

  return AllocationResult(
    allocations: allocations,
    allocatedBeforeDiscretionary: Money.sum(
      allocations
          .where((a) => a.priority != Priority.p9Discretionary)
          .map((a) => a.allocated),
      currency,
    ),
    // Summed per claim rather than by subtracting one mandatory total from
    // total liquidity, because P6 sits above P7: a funded buffer can absorb
    // liquidity a hard goal then cannot reach, and a single subtraction would
    // report no gap while a hard commitment is underfunded (§13, T36).
    mandatoryFundingGap: shortfallWhere((a) => a.priority.isMandatory),
    bufferShortfall: shortfallWhere((a) => a.priority == Priority.p6Buffer),
    flexibleShortfall: shortfallWhere((a) => a.priority == Priority.p8Flexible),
    protectionHorizonEnd: protectionHorizonEnd,
    extendedBeyondDecisionHorizon: extended,
  );
}
